### PR TITLE
🧹Clean (and upgrade) application.js

### PR DESCRIPTION
**Pull request by josemigallas**
This file contains many outdated js snippets that are dragging us from upgrading jQuery and also makes it very hard to maintain.

I've tried to break every single callback / listener and I've found out most of them refer to ids or selectors long ago removed from 3scale.

The ones that are still used have been upgraded to jQuery 3.x as much as possible.

### CODE REVIEW

File: app/javascript/packs/fields_definitions_form.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    $hidden[0].checked = false
+    $hidden.attr('disabled', 'true')
+
+    const $readonly = $<HTMLInputElement>('#fields_definition_read_only')
+    $readonly[0].checked = false
+    $readonly.attr('disabled', 'true')
+  }
+
+  //edit view
+  if (document.getElementById('fields-definitions-edit-view') !== null) {
```
Comment (mayorova) [inline app/javascript/packs/fields_definitions_form.ts:89]: Just curious, why is this one using JavaScript, and not jQuery?

File: app/views/plans/_selector_widget.html.erb [hunk 2]
```diff
@@ -1,3 +1,6 @@
+<%# DELETEME !!! %>
+<% raise RuntimeError, 'If this does not throw, then this template is not rendered' %>
```
Comment (mayorova) [inline app/views/plans/_selector_widget.html.erb:2]: I guess this is for tests, right? so, we can delete the template then? :grimacing:

Comment (josemigallas) [reply app/views/plans/_selector_widget.html.erb:2]: 🙉 gosh!

File: app/javascript/packs/fields_definitions_form.ts (hunk 1 above)
Comment (josemigallas) [reply app/javascript/packs/fields_definitions_form.ts:89]: I will use jQuery only when strictly necessary to reduce dependency to a minimum.

File: app/javascript/src/Common/application.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+const jQuery1 = window.$
+
+export default function (): void {
+  // disable links with 'data-disabled' attribute and display alert instead
+  // delegation on body fires before rails.js. FIXME: this is not a valid guard. If "data-disabled"
+  // removed manually from the DOM, the request will still go through. The link should be disabled
+  // (not clickable) and the server should return a flash error.
+  $('body').on('click', 'a[data-disabled]', ({ currentTarget, stopImmediatePropagation }) => {
+    stopImmediatePropagation()
```
Comment (mayorova) [inline app/javascript/src/Common/application.ts:11]: The original code also had `preventDefault`. So, it's not needed?

Comment (josemigallas) [reply app/javascript/src/Common/application.ts:11]: It's already covered by `return false`. HOWEVER, after further investigation it turns out returning a value was deprecated and is recommended not to use it (even though most browsers supports it). So I will revert it.

File: app/javascript/src/Common/application.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+const jQuery1 = window.$
+
+export default function (): void {
+  // disable links with 'data-disabled' attribute and display alert instead
+  // delegation on body fires before rails.js. FIXME: this is not a valid guard. If "data-disabled"
+  // removed manually from the DOM, the request will still go through. The link should be disabled
+  // (not clickable) and the server should return a flash error.
+  $('body').on('click', 'a[data-disabled]', (event) => {
+    event.stopPropagation()
+    event.preventDefault()
```
Comment (josemigallas) [inline app/javascript/src/Common/application.ts:12]: This used to be `return false`.

File: app/javascript/src/Common/application.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    jQuery1(currentTarget as HTMLAnchorElement).colorbox({ open: true })
+    preventDefault()
+  })
+
+  // TODO: replace .fancybox with .colorbox
+  // This is used in some modals with a "Cancel" button.
+  $(document).on('click', '.fancybox-close', ({ preventDefault, stopPropagation }) => {
+    jQuery1.colorbox.close()
+    preventDefault()
+    stopPropagation()
```
Comment (josemigallas) [inline app/javascript/src/Common/application.ts:48]: This used to be return false.

File: app/javascript/src/Common/application.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+export default function (): void {
+  // disable links with 'data-disabled' attribute and display alert instead
+  // delegation on body fires before rails.js. FIXME: this is not a valid guard. If "data-disabled"
+  // removed manually from the DOM, the request will still go through. The link should be disabled
+  // (not clickable) and the server should return a flash error.
+  $('body').on('click', 'a[data-disabled]', (event) => {
+    event.stopPropagation()
+    event.preventDefault()
+    event.stopImmediatePropagation()
```
Comment (josemigallas) [inline app/javascript/src/Common/application.ts:13]:
```diff
-    return false;
+    event.stopPropagation()
+    event.preventDefault()
    event.stopImmediatePropagation()
```

File: app/javascript/src/Common/application.ts (hunk 5 above)
Comment (josemigallas) [inline app/javascript/src/Common/application.ts:48]:
```diff
-    return false;
+    event.stopPropagation()
+    event.preventDefault()
```

File: app/javascript/src/Common/application.ts [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+   * responseText inside the XHR response. MUST USE RAILS-JQUERY since ajax:error is triggered by
+   * rails (do not use $ from node_modules).
+   *
+   * DEPRECATED: replace jquery/colorbox modals with Patternfly modals.
+   * TODO: add a colorbox specific selector to separate this with non-legacy implementations.
+   */
+  jQuery1(document).on('ajax:error', 'form:not(.pf-c-form)', (event, xhr: JQueryXHR, status) => {
+    if (status === 'error') {
+      jQuery1.colorbox({ html: xhr.responseText })
+      event.stopPropagation()
```
Comment (josemigallas) [inline app/javascript/src/Common/application.ts:62]: Stays the same

File: app/javascript/src/Common/application.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+const jQuery1 = window.$
+
+export default function (): void {
+  // disable links with 'data-disabled' attribute and display alert instead
+  // delegation on body fires before rails.js. FIXME: this is not a valid guard. If "data-disabled"
+  // removed manually from the DOM, the request will still go through. The link should be disabled
+  // (not clickable) and the server should return a flash error.
+  $('body').on('click', 'a[data-disabled]', (event) => {
+    event.stopPropagation()
```
Comment (mayorova) [inline app/javascript/src/Common/application.ts:11]:
```suggestion
```

I understand that `stopImmediatePropagation` kinda includes `stopPropagation` (it does the same and more), so it's redundant to put both.

Comment (josemigallas) [reply app/javascript/src/Common/application.ts:11]: My mistake 👍🏻

**Review by mayorova — APPROVED** [review]
I haven't retested after the changes, but as the commits have been nicely pushed without rebasing, it's easy to check.

(except that the tests fail, so probably something has been broken while refactoring :upside_down_face: )

**Pull request merged**